test/engine/update.test.lua

```diff
@@ -160,3 +160,78 @@ t:update({{'+', '[1]', 50}})
 s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
 
 s:drop()
+
+--
+-- gh-3378: allow update absent nullable fields
+
+-- '!'
+s = box.schema.create_space('test')
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', 4, 0}})
+s:drop()
+
+-- Update respects field_count
+s = box.schema.create_space('test', {field_count = 2})
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', 3, 0}})
+s:update({1}, {{'=', 3, 0}})
+s:drop()
+
+-- '='
+s = box.schema.create_space('test')
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'=', 4, 0}})
+s:drop()
+
+-- Ops that must still fail
+format = {}
+format[1] = {name = 'field1', type = 'unsigned'}
+format[2] = {name = 'field2', type = 'unsigned'}
+format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
+s = box.schema.create_space('test', {format = format})
+pk = s:create_index('pk')
+s:replace{1, 2, box.NULL}
+s:update({1}, {{'+', 3, 0}})
+s:update({1}, {{'-', 3, 0}})
+s:update({1}, {{'&', 3, 0}})
+s:update({1}, {{'|', 3, 0}})
+s:update({1}, {{'^', 3, 0}})
+s:update({1}, {{':', 3, 1, 0, '!!'}})
+s:drop()
+
+-- Negative field number, fixed field_count
+format = {}
+format[1] = {name = 'field1', type = 'unsigned'}
+format[2] = {name = 'field2', type = 'unsigned'}
+format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
+format[4] = {name = 'field4', type = 'unsigned', is_nullable = true}
+s = box.schema.create_space('test', {format = format, field_count = 4})
+pk = s:create_index('pk')
+s:replace{1, 2, box.NULL, box.NULL}
+s:update({1}, {{'!', -1, 42}})
+s:update({1}, {{'=', -1, 128}})
+s:drop()
+
+-- Negative field number, no field_count
+format = {}
+format[1] = {name = 'field1', type = 'unsigned'}
+format[2] = {name = 'field2', type = 'unsigned'}
+format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
+format[4] = {name = 'field4', type = 'unsigned', is_nullable = true}
+s = box.schema.create_space('test', {format = format})
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', -1, 42}})
+s:update({1}, {{'=', -1, 128}})
+s:drop()
+
+-- '#' doesn't trim nulls
+s = box.schema.create_space('test')
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', 4, 0}})
+s:update({1}, {{'#', 4, 1}})
+s:drop()
```

Comment:
Please, add more tests with json paths. What if I use `!` and `=` on a field whose parent's index is out of range? For example `[5][1] = 100`, where field `[5]` does not exist. The same with maps (`[5].key`).

Reply:
Agree. Even behavior isn't considered to be expected (I mean like @Gerold103  suggest to do right now), test on the subj would be nice to see.